### PR TITLE
Add Helper Fn for AWS Secrets Manager

### PR DESCRIPTION
**Pull request by amy17519**
`library(botor)` currently has `ssm_get_parameter`, which can retrieve from both AWS Parameter Store and Secrets Manager (if you know all the details to make that happen)

I wrote two helper fns `ssm_get_secrets()` and `sm_get_secrets()` for retrieving secrets in AWS Secrets Manager. Users can specify their secrets in two formats:
xx_get_secrets(path = 'your/secret/key')
xx_get_secrets(path = 'your/secret', key = key)

1. `ssm_get_secrets()` uses `botocore.client.SSM`

Complexities:
- Users have to grant permission to both Parameter Store and SecretManager, which is fine when the users use both. It'd be hard to debug when the users use only SecretManager
- Users have to remember to enter key in the format of `/aws/reference/secretsmanager/your/secret`. No big deal and we can certainly make this clear in description.

2. `sm_get_secrets()` uses `botocore.client.secretsmanager`. Function and AWS permission would be cleaner. The only downside is that we need to add more functions 

@daroczig 

UPDATE (May 21, 2020):
Decide to only add `sm_get_secrets()` for retrieving secrets from AWS Secret Manager. The function supports retrieving value from json-structured secrets by a key or a vector of keys. Users can also get secrets in raw format (useful when secrets are not json or you simply want unparsed output)

### CODE REVIEW

File: R/ssm.R [hunk 1]
```diff
@@ -18,3 +18,69 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+#' Read AWS System Manager's Secrets Manager via Parameter Store
+#' @param path name/path of the key to be read
+#' @param key keyID
+#' @param decrypt decrypt the value or return the raw ciphertext
+#' @importFrom checkmate assert_string
+#' @importFrom jsonlite fromJSON
+#' @return (optionally decrypted) value
+#' @export
+ssm_get_secrets <- function(path, key = NULL, decrypt = TRUE) {
+    # Usage: path = your/secretname/keyID
```
Comment (amy17519) [inline R/ssm.R]: TODO: add to fn help doc

File: R/ssm.R [hunk 2]
```diff
@@ -18,3 +18,69 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+    # or 
+    # path = your/seretname, key = keyID
```
Comment (amy17519) [inline R/ssm.R]: TODO: add to fn help doc

File: R/ssm.R [hunk 3]
```diff
@@ -18,3 +18,69 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+        path <- paste(parts[1:length(parts) - 1], collapse = "/")
+    }
+
+    log_trace("Looking up keyID %s from SecretId %s in AWS Secrets Manager", key, path)
+    
+    resp <- trypy(ssm()$get_parameter(
+                Name = path,
+                WithDecryption = decrypt)$Parameter$Value)
+
+    fromJSON(resp)[[key]]
```
Comment (daroczig) [inline R/ssm.R]: Although the SM UI enforces the user to provide JSON, but it stores that as a standard string ... and no need to pass a JSON eg via the API (eg it also supports binary). So maybe let's make this optional?

File: R/ssm.R [hunk 4]
```diff
@@ -18,3 +18,69 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+#' @return (optionally decrypted) value
+#' @export
+ssm_get_secrets <- function(path, key = NULL, decrypt = TRUE) {
+    # Usage: path = your/secretname/keyID
+    # or 
+    # path = your/seretname, key = keyID
+    
+    assert_string(path, pattern = "^/aws/reference/secretsmanager/")
+
+    if (is.null(key)) {
```
Comment (daroczig) [inline R/ssm.R]: Awesome, thanks -- this is a nice feature! But we might need to wrap it in a try loop .. to see how many levels we might need to go up/down?

File: R/ssm.R [hunk 5]
```diff
@@ -18,3 +18,69 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+#' @importFrom checkmate assert_string
+#' @importFrom jsonlite fromJSON
+#' @return (optionally decrypted) value
+#' @export
+sm_get_secrets <- function(path, key = NULL) {
+    # Format: path = your/secretname/key 
+    # or 
+    # path = your/seretname, key = key
+    
+    if (is.null(key)) {
```
Comment (daroczig) [inline R/ssm.R]: same as above and maybe we can DRY this out a bit?

**Comment by daroczig** [comment]
This is really cool, @amy17519  -- thanks a lot :bowing_man: 

Added some comments above, please let's discuss before merging.

File: R/ssm.R [hunk 6]
```diff
@@ -18,3 +18,76 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+    
+    resp <- trypy(ssm()$get_parameter(
+                Name = path,
+                WithDecryption = decrypt)$Parameter$Value)
+
+    fromJSON(resp)[[key]]
+}
+
+
+#' Read AWS System Manager's Secrets Manager via Secret Manager
```
Comment (amy17519) [inline R/ssm.R]: how does this look? @daroczig

File: R/ssm.R [hunk 7]
```diff
@@ -18,3 +18,76 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
         Name = path,
         WithDecryption = decrypt)$Parameter$Value)
 }
+
+#' The default, fork-safe AWS Systems Manager (SecretManager) client on the top of \code{botor}
```
Comment (daroczig) [inline R/ssm.R]: can you pls move `sm` and `sm_get_secrets` to their own `sm.R` file?

File: R/ssm.R [hunk 8]
```diff
@@ -18,3 +18,76 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+
+#' Read AWS System Manager's Secrets Manager via Secret Manager
+#' @param path name/path of the key to be read
+#' @param keyinpath logical. Default TRUE
+#' @param output_format return format. Either as raw_string or value
+#' @importFrom checkmate assert_string assert_logical
+#' @importFrom jsonlite fromJSON
+#' @return (optionally decrypted) value
+#' @export
+sm_get_secrets <- function(path, keyinpath = TRUE, output_format = c("raw_string", "value")) {
```
Comment (daroczig) [inline R/ssm.R]: maybe name to `sm_get_secret`? not sure why plural

File: R/ssm.R [hunk 9]
```diff
@@ -18,3 +18,76 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+
+#' Read AWS System Manager's Secrets Manager via Parameter Store
+#' @param path name/path of the key to be read
+#' @param key keyID
+#' @param decrypt decrypt the value or return the raw ciphertext
+#' @importFrom checkmate assert_string
+#' @importFrom jsonlite fromJSON
+#' @return (optionally decrypted) value
+#' @export
+ssm_get_secrets <- function(path, key = NULL, decrypt = TRUE) {
```
Comment (daroczig) [inline R/ssm.R]: I think now that you wrote `sm_get_secret`, we can just leave this out?

File: R/ssm.R [hunk 10]
```diff
@@ -18,3 +18,76 @@ ssm_get_parameter <- function(path, decrypt = TRUE) {
...
+#' @param output_format return format. Either as raw_string or value
+#' @importFrom checkmate assert_string assert_logical
+#' @importFrom jsonlite fromJSON
+#' @return (optionally decrypted) value
+#' @export
+sm_get_secrets <- function(path, keyinpath = TRUE, output_format = c("raw_string", "value")) {
+    match.arg(output_format)
+    assert_logical(keyinpath)
+    
+    if (keyinpath) {
```
Comment (daroczig) [inline R/ssm.R]: sorry if the above comments were not clear, trying to get an actual example -- let's say we store the below JSON under the `foo` key in SM:

```
{"bar":{"baz":42,"bay":24}} 
```

Then I get `42` from this by requesting `foo/bar/baz`

Comment (daroczig) [reply R/ssm.R]: So we might try to look up `foo/bar/baz` first, and return if succeeds, but if not, then try `foo/bar` and then look up its `baz` piece ... if failing, still looking up `foo` in SM and then looking up its `bar` element and then going deeper for `baz`.

Thoughts?

File: R/sm.R [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+#' @param path name/path of the key to be read
+#' @param key single key or a vector of keys. 
+#' @param parse_json logical. Default TRUE
+#' @importFrom checkmate assert_vector assert_logical
+#' @importFrom jsonlite fromJSON
+#' @importFrom logger log_warn
+#' @return (optionally decrypted) value
+#' @export
+sm_get_secret <- function(path, key = NULL, parse_json = TRUE) {
+  assert_logical(parse_json)
```
Comment (daroczig) [inline R/sm.R]:
```suggestion
  assert_flag(parse_json)
```

File: R/sm.R [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  assert_vector(key, null.ok = TRUE)
+
+  if (!is.null(key) & parse_json == FALSE) {
```
Comment (daroczig) [inline R/sm.R]: we can stop early with these, eg if the `key` is `NULL`, no need to check the value of `parse_json`:

```suggestion
  if (!is.null(key) && parse_json == FALSE) {
```

File: R/sm.R [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    stop('Key(s) must be provided if parse_json = TRUE')
+  }
+
+  log_trace("Looking up SecretId %s in AWS Secrets Manager", path)
+
+  resp <- trypy(sm()$get_secret_value(
+    SecretId = path))$SecretString
```
Comment (daroczig) [inline R/sm.R]:
```suggestion
  resp <- trypy(sm()$get_secret_value(SecretId = path))$SecretString
```

File: R/sm.R [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+  if (parse_json) {
+    resp <- fromJSON(resp)[[key]]
+    if (is.null(resp)) log_warn("Key %s does not exist", key)
+    resp
```
Comment (daroczig) [inline R/sm.R]: I think it's redundant to return `resp` here and on L37 as well, it's enough to write this once after the `if` condition?

File: R/sm.R [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+#' @importFrom jsonlite fromJSON
+#' @importFrom logger log_warn
+#' @return (optionally decrypted) value
+#' @export
+sm_get_secret <- function(path, key = NULL, parse_json = TRUE) {
+  assert_flag(parse_json)
+  assert_vector(key, null.ok = TRUE)
+
+  if (!is.null(key) && parse_json == FALSE) {
+    stop('Key(s) must be provided if parse_json = TRUE')
```
Comment (daroczig) [inline R/sm.R]: I think this is actually the other way around ... you can ask to parse JSON even without a key, but not extract a key without parsing JSON:

```suggestion
    stop('Need to set parse_json=TRUE when extracting keys')
```

File: R/sm.R [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  log_trace("Looking up SecretId %s in AWS Secrets Manager", path)
+
+  resp <- trypy(sm()$get_secret_value(SecretId = path))$SecretString
+
+  if (parse_json) {
+    resp <- fromJSON(resp)[[key]]
+    if (is.null(resp)) log_warn("Key %s does not exist", key)
+  }
```
Comment (daroczig) [inline R/sm.R]: As per above, it's possible that we just parse JSON but there are no `key` extract:

```suggestion
  if (parse_json) {
    resp <- fromJSON(resp)
  }
  for (k in key) {
    resp <- resp[[k]]
  }
```

Feel free to skip the loop for your `resp[[key]]` solution (this just looked more readable and a bit shorter to me).

**Review by daroczig — APPROVED** [review]
Thanks a lot, @amy17519 :+1: :bow: 

Any thoughts on the above suggestions?

If looks good, please accept those and we can merge, otherwise let's chat :)

**Pull request merged**